website/static/js/registrationModal.js

```diff
@@ -19,12 +23,162 @@ var MAKE_EMBARGO = {
     value: 'embargo',
     message: 'Enter registration into embargo'
 };
+var MESSAGES = {
+    selectTitle: 'Select components',
+    confirmTitle: 'Confirm',
+    embargoTitle: 'Embargo',
+    selectNodes: 'Select the components you would like to register.',
+};
 
-var RegistrationViewModel = function(confirm, prompts, validator) {
+function _flattenNodeTree(nodeTree) {
+    var ret = [];
+    var stack = [nodeTree];
+    while (stack.length) {
+        var node = stack.pop();
+        if(node.children){
+            $.each(node.children, function(_, child){
+                child.parent = node.node.id;
+            });
+            stack = stack.concat(node.children);
+        }
+        ret.push(node);
 
-    var self = this;
+    }
+    return ret;
+}
+
+/**
+ * take treebeard tree structure of nodes and get a dictionary of parent node and all its
+ * children
+ */
+function getNodesOriginal(nodeTree, nodesOriginal) {
+    var flatNodes = _flattenNodeTree(nodeTree);
+    $.each(flatNodes, function(_, nodeMeta) {
+        nodesOriginal[nodeMeta.node.id] = {
+            selected: false,
+            id: nodeMeta.node.id,
+            title: nodeMeta.node.title,
+            isAdmin: nodeMeta.node.is_admin,
+            changed: false,
+            parent: nodeMeta.parent
+        };
+    });
+    nodesOriginal[nodeTree.node.id].isRoot = true;
+    nodesOriginal[nodeTree.node.id].selected = true;
+    nodesOriginal[nodeTree.node.id].disabled = true; // The user must register the root.
```

Comment:
This seems to go against the line of the acceptance criteria: "Admins on components can register a component and its children (without registering top-level project) [current behavior]"

I can see how to achieve this, you go to the component and then register from there -- but was curious if how easy it would be to let an admin register just pieces of the registration's components from the top level if all other criteria are met?

Reply:
I see how you could think that, but the "root" here is just the node the user is registering from, not necessarily the root of the whole thing. We're assume the component the user is registering is the "root" of the registration. I think this makes it clear what registration draft stands for what, but I should probably double check with Product.

Reply:
@caseyrollins Did we talk about this during the demo? I have a fuzzy memory of this getting brought up.